src/Compilers/CSharp/Test/Symbol/Symbols/CovariantReturnTests.cs

```diff
@@ -1659,11 +2115,793 @@ End Class
+public class Base
+{
+    public virtual object M() => null;
+}
+";
+            var ref0 = CreateCompilationWithoutCovariantReturns(s0).EmitToImageReference();
+
+            var s1a = @"
+public class Mid : Base
+{
+}
+";
+            var ref1a = CreateCompilationWithoutCovariantReturns(s1a, references: new[] { ref0 }, assemblyName: "ref1").EmitToImageReference();
+
+            var s1b = @"
+public class Mid : Base
+{
+    public override object M() => null;
+}
+";
+            var ref1b = CreateCompilationWithCovariantReturns(s1b, references: new[] { ref0 }, assemblyName: "ref1").EmitToImageReference();
+
+            var s2 = @"
+public class Derived : Mid
+{
+    public override string M() => null;
+}
+";
+            var comp = CreateCompilationWithoutCovariantReturns(s2, references: new[] { ref0, ref1a }).VerifyDiagnostics(
+                // (4,28): error CS8652: The feature 'covariant returns' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public override string M() => null;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "M").WithArguments("covariant returns").WithLocation(4, 28)
+                );
+            verify1(comp);
+
+            comp = CreateCompilationWithCovariantReturns(s2, references: new[] { ref0, ref1a }).VerifyDiagnostics(
+                );
+            verify1(comp);
+            verify1(CompilationReferenceView(comp, ref0, ref1a));
+            verify1(MetadataView(comp, ref0, ref1a));
+            verify1(RetargetedView(comp, ref0, ref1a));
+
+            verify2(CompilationReferenceView(comp, ref0, ref1b));
+            verify2(MetadataView(comp, ref0, ref1b));
+            verify2(RetargetedView(comp, ref0, ref1b));
+
+            static void verify1(CSharpCompilation comp)
+            {
+                VerifyOverride(comp, "Derived.M", "System.String Derived.M()", "System.Object Base.M()");
+            }
+
+            static void verify2(CSharpCompilation comp)
+            {
+                // When viewed from metadata, we do not see a relationship between Derived.M and Mid.M because
+                // there is nothing in the metadata to indicate there is relationship. The compiler
+                // does no simulate the covariant language rules on metadata nor does the compiler simulate the
+                // virtual slot unification that the runtime does.
+                VerifyOverride(comp, "Derived.M", "System.String Derived.M()", "System.Object Base.M()");
+                VerifyOverride(comp, "Mid.M", "System.Object Mid.M()", "System.Object Base.M()");
+            }
+        }
+
+        [Fact]
+        public void OverlappingMethodimplRequirements_01()
+        {
+            var source = @"
+public class A
+{
+    public virtual object get_P() => null;
+}
+
+public class B : A
+{
+    public virtual object P => null;
+}
+
+public class C : B
+{
+    public override string get_P() => null;
```

Comment:
>string [](start = 20, length = 6)

I think it would be interesting to test what happens when both overrides use object as return types, including checking what runtime "thinks" it should do.

Reply:
I will add a runtime expected output to the existing test `MethodAccessorMethodOverridingRoundTrip` in `CodeGenOverridingAndHiding`.

---
In reply to: [424784704](https://github.com/dotnet/roslyn/pull/44025#discussion_r424784704) [](ancestors = 424784704)